### DISCUSSION TITLE
Problemi con Caddy e reverse proxy

### DISCUSSION THREAD

**Discussion by vpomax** [indomus/forum#788] [HUB: Home Assistant]
La mia configurazione hardware consiste in un modem Fritzbox 7590 (192.168.0.1) collegato alla linea internet (con port forwarding sulle porte 80, 8123, 443 e 9000)... collegato in cascata con una router ASUS AC68U (192.168.1.1) con il port forwarding sulle stesse porte (tranne che la porta 9000 verso la 8123) e collegato a questo un NUC con Proxmox e Home Assistant HASSIO (192.168.1.2).

Purtroppo non riesco a risolvere il problema del reverse proxy, prima avevo installato NGINX... ma anche adesso con Caddy2.

Ho seguito questa guida : https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-via-reverse-proxy/

Questo è il Caddyfile:

(https_header) {
  header {
    Strict-Transport-Security "max-age=31536000; includeSubdomains"
    X-XSS-Protection "1; mode=block"
    X-Content-Type-Options "nosniff"
    X-Frame-Options "SAMEORIGIN"
    Referrer-Policy "same-origin"
  }
}

https://miodominio.duckdns.org:9000 {
  import https_header
  reverse_proxy http://192.168.1.2:8123
}

Questo è il log del Caddy2:

[cont-init.d] 00-banner.sh: exited 0.
[cont-init.d] 01-log-level.sh: executing... 
Log level is set to INFO
[cont-init.d] 01-log-level.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
INFO: Prepare Caddy...
INFO: Use built-in Caddy
v2.4.5 h1:P1mRs6V2cMcagSPn+NWpD+OEYUYLIf6ecOa48cFGeUg=
INFO: Prepare Caddyfile...
INFO: Caddyfile found at /config/caddy/Caddyfile
INFO: Run Caddy...
{"level":"info","ts":1632778589.261656,"msg":"using provided configuration","config_file":"/config/caddy/Caddyfile","config_adapter":""}
{"level":"warn","ts":1632778589.2627037,"msg":"input is not formatted with 'caddy fmt'","adapter":"caddyfile","file":"/config/caddy/Caddyfile","line":2}
{"level":"info","ts":1632778589.272741,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["127.0.0.1:2019","localhost:2019","[::1]:2019"]}
{"level":"info","ts":1632778589.274262,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
{"level":"info","ts":1632778589.2756386,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc0002c1a40"}
{"level":"info","ts":1632778589.2855363,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["miodominio.duckdns.org"]}
{"level":"info","ts":1632778589.2861092,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/ssl/caddy"}
{"level":"info","ts":1632778589.299297,"msg":"autosaved config (load with --resume flag)","file":"/data/caddy/autosave.json"}
{"level":"info","ts":1632778589.2995071,"msg":"serving initial configuration"}
{"level":"info","ts":1632778589.2997956,"logger":"watcher","msg":"watching config file for changes","config_file":"/config/caddy/Caddyfile"}
{"level":"info","ts":1632778589.3003664,"logger":"tls","msg":"finished cleaning storage units"}

se cerco di accedere all'indirizzo remoto https://miodominio.duckdns.org carica la pagina di HA  scrive initializing... e si blocca.

**Comment by Scialla** [comment]
io ci capisco poco ma, o aggiungi la porta 9000 quando cerchi di accedere, o aggiungi in configurazione il reverse proxy anche dell'fqdn senza porta....
difatti , visto che hai lasciato il tuo fqdn, ho provato e lo raggiungo mettendo la porta, però collegandomi in http e non in https, quindi hai pure problemi coi certificati.

PS togli il tuo fqdn

**Reply by indomus** [reply]
come dice Scialla, o punti a https://miodominio.duckdns.org:9000 o non ti collegherai mai (sempre che tu abbia configurato correttamente anche il port forwarding, chiaro)

**Reply by snakuzzo** [reply]
Mamma mia @vpomax ...ma che brutta architettura!!! :D ... come complicarsi la vita insomma. Perché usare due router? Non puoi buttare via il router e/o configurare il fritz in bridge ed usare solo il router ?

Ad ogni modo devi fare in modo che da fuori sia raggingibile Caddy sulla 443 e/o sulla 80 per le challenge che ti occorrono per la richiesta e rinnovo dei certificati. Stesso dicasi per HA.

Innanzi tutto la guida che hai linkato non è per HassOS (e non Hassio come hai scritto), anche se di base la logica è la stessa. [QUESTA](https://indomus.it/guide/installare-e-configurare-caddy-su-home-assistant-os-hassos-hassio/) è la guida per installarlo come addon su HassOS se dovesse servirti.

Poi noto che ci sono idee moooolto confuse sul reverse proxy. Dico questo perché vedo che hai fatto copia-incolla dalla guida scrivendo che HA, che risponde localmente sulla 8123, viene proxato da caddy che risponde sulla porta 9000...e poi tenti di contattarlo sulla 443 :D... che senso avrebbe?
Ora, non so dove hai installato Caddy, ma dando per scontato che sia tutto sullo stesso server e che tu voglia esporre HA sulla porta 9000 come da Caddyfile, il giro deve essere...

INTERNET -> modem 443 -> router 443 -> proxmox 443
INTERNET -> modem 80 -> router 80 -> proxmox 80
INTERNET -> modem 9000 -> router 9000 -> proxmox 9000 -> Caddy 9000 -> HA 8123

P.S. Non date per scontato che una guida valga per qualsiasi tipo di configurazione. La guida serve anche come spunto e non va fatto copia-incolla senza capire. Quindi ti consiglio di fermarti, cancellare tutto e ripartire da zero cercando di comprendere cosa stai facendo

**Reply by vpomax** [reply]
il doppio router era dovuto al fatto che fino a qualche giorno fa avevo come modem router un orribile Fastgate che non mi permetteva di fare molto anzi ad un certo punto si era anche bloccato e qualsiasi parametro cambiavo (es. un port forwarding) mi diceva di contattare l'assistenza. Ecco il motivo per cui ho comprato il Fritzbox 7590. Siccome tutta la configurazione di rete attualmente ce l'ho sull'ASUS comprese tutte le periferiche wifi della domotica (30 devices).... volevo fare il passaggio al Fritzbox in due fasi... ma evidentemente devo cambiare programma...

Comunque attualmente ho...
Sul modem Fritzbox il port forwarding 
- dalla porta 80 alla 80 dell'Asus
- dalla porta 443 alla 443 dell'Asus
- dalla porta 9000 alla 9000 dell'Asus 

Sul router Asus il port forwarding
- dalla porta 80 alla 80 di Proxmox (NUC)
- dalla porta 443 alla 443 di Proxmox (NUC)
- dalla porta 9000 alla 9000 di Proxmox (NUC)

**Reply by snakuzzo** [reply]
ok...se non hai scritto cavolate...ha deve rispondere su

`https://tuodominio:9000`

Non vedo traccia di challenge nel log di Caddy, quindi quello dovrebbe essere stato rilasciato correttamente

**Reply by vpomax** [reply]
si riesco anch'io ad accedere specificando la porta 9000
questo però significa che Caddy2 non funziona perché il Caddyfile dovrebbe fare il redirect dalla 9000 ala 8123

**Reply by snakuzzo** [reply]
Scusami...
ma Caddy non fa una redirect, fa da proxy.
Home Assistant da fuori lo "vedi" sulla porta 9000... punto.
Se questo è il comportamento che riscontri è esattamente quello che deve fare.

**Reply by vpomax** [reply]
Ahh.. ok... quindi è corretto così? bene. 
Pensavo che bastasse digitare l'indirizzo senza specificare la porta 9000
Grazie per il supporto.

**Comment by manny199** [comment]
@snakuzzo 
Ho qualche problema con una nuova installazione di caddy su docker compose.
Purtroppo non genera i certificati ed ottengo il seguente errore:
```
{"level":"info","ts":1644678550.8602955,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"}
{"level":"warn","ts":1644678550.8647745,"msg":"input is not formatted with 'caddy fmt'","adapter":"caddyfile","file":"/etc/caddy/Caddyfile","line":2}
{"level":"info","ts":1644678550.8693566,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]}
{"level":"info","ts":1644678550.8698287,"logger":"http","msg":"server is listening only on the HTTPS port but has no TLS connection policies; adding one to enable TLS","server_name":"srv0","https_port":443}
{"level":"info","ts":1644678550.869884,"logger":"http","msg":"enabling automatic HTTP->HTTPS redirects","server_name":"srv0"}
{"level":"info","ts":1644678550.8701181,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x3210140"}
{"level":"info","ts":1644678550.870731,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/data/caddy"}
{"level":"info","ts":1644678550.8709505,"logger":"http","msg":"enabling automatic TLS certificate management","domains":["morinello.duckdns.org"]}
{"level":"info","ts":1644678550.8711748,"logger":"tls","msg":"finished cleaning storage units"}
{"level":"info","ts":1644678550.8714359,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}
{"level":"info","ts":1644678550.8714733,"msg":"serving initial configuration"}
{"level":"info","ts":1644678550.8724453,"logger":"tls.obtain","msg":"acquiring lock","identifier":"morinello.duckdns.org"}
{"level":"info","ts":1644678550.8794484,"logger":"tls.obtain","msg":"lock acquired","identifier":"morinello.duckdns.org"}
{"level":"info","ts":1644678550.8864791,"logger":"tls.issuance.acme","msg":"waiting on internal rate limiter","identifiers":["morinello.duckdns.org"],"ca":"https://acme-staging-v02.api.letsencrypt.org/directory","account":""}
{"level":"info","ts":1644678550.8865376,"logger":"tls.issuance.acme","msg":"done waiting on internal rate limiter","identifiers":["morinello.duckdns.org"],"ca":"https://acme-staging-v02.api.letsencrypt.org/directory","account":""}
{"level":"info","ts":1644678552.1482081,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"morinello.duckdns.org","challenge_type":"http-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
{"level":"error","ts":1644678553.2088466,"logger":"tls.issuance.acme.acme_client","msg":"challenge failed","identifier":"morinello.duckdns.org","challenge_type":"http-01","problem":{"type":"urn:ietf:params:acme:error:unauthorized","title":"","detail":"Invalid response from http://morinello.duckdns.org/.well-known/acme-challenge/WmNQk1QmhW4e1Pdt_7WH8T8uWT2JhDaFJYBFWwBCRSw [46.169.28.19]: \"<!DOCTYPE HTML PUBLIC \\\"-//W3C//DTD HTML 4.01//EN\\\" \\\"http://www.w3.org/TR/html4/strict.dtd\\\">\\r\\n<HTML><HEAD><TITLE>Impossibile trova\"","instance":"","subproblems":[]}}
{"level":"error","ts":1644678553.2090018,"logger":"tls.issuance.acme.acme_client","msg":"validating authorization","identifier":"morinello.duckdns.org","problem":{"type":"urn:ietf:params:acme:error:unauthorized","title":"","detail":"Invalid response from http://morinello.duckdns.org/.well-known/acme-challenge/WnNQs1QbbW4e1Paa_7WH8T8uWT2JnniFJYBFWwBCRSw [37.102.23.12]: \"<!DOCTYPE HTML PUBLIC \\\"-//W3C//DTD HTML 4.01//EN\\\" \\\"http://www.w3.org/TR/html4/strict.dtd\\\">\\r\\n<HTML><HEAD><TITLE>Impossibile trova\"","instance":"","subproblems":[]},"order":"https://acme-staging-v02.api.letsencrypt.org/acme/order/43778818/4477565408","attempt":1,"max_attempts":3}
{"level":"info","ts":1644678554.592097,"logger":"tls.issuance.acme.acme_client","msg":"trying to solve challenge","identifier":"morinello.duckdns.org","challenge_type":"tls-alpn-01","ca":"https://acme-staging-v02.api.letsencrypt.org/directory"}
```
il mio caddyflie per la generazione iniziale dei certificati, che poi dovrà essere modificato è:
```
{
    acme_ca https://acme-staging-v02.api.letsencrypt.org/directory 
}

https://tronxuff.duckdns.org { 
  respond "prova_test"
  file_server
}
```
mentre il mio file su docker-compose è:
```
  caddy:
    container_name: caddy
    image: caddy:latest
    volumes:
    - "/home/pi/caddy/Caddyfile:/etc/caddy/Caddyfile"
    - "/home/pi/caddy/data:/data"
    environment:
    - "TZ=Europe/Rome"
    restart: always
    ports:
    - "80:80"
    - "443:443"
    - "9000:9000"
```
In sostanza non mi crea la cartella "certificates" dentro caddy

**Reply by snakuzzo** [reply]
Assicurati di aver aperto correttamente le porte
